keps/sig-cloud-provider/1959-service-lb-class-annotation/README.md

```diff
@@ -0,0 +1,203 @@
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [ServiceClass Resource](#serviceclass-resource)
+  - [Provider-Specific Annotations](#provider-specific-annotations)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+When Service Type=LoadBalancer is enabled by a Kubernetes cloud provider, it is a global
+configuration that applies for all Service Type=LoadBalancer resources in a given cluster.
+This becomes problematic if users want to leverage multiple Service Type=LoadBalancer
+implementations in a given cluster.
+
+The new [Services APIs](https://github.com/kubernetes-sigs/service-apis) addresses this already
+with the GatewayClass resource. However, until Gateway/GatewayClass APIs become mature, we should
+support similar functionality for Services of Type=LoadBalancer. Introducing a new resource like
+`ServiceClass` is probably not worthwhile given that there are new APIs already in development.
+This KEP proposes a light-weight approach for Service Type=LoadBalancer by introducing a Service
+annotation `service.kubernetes.io/load-balancer-class`.
+
+## Motivation
+
+The main use-case for this feature is being able to support multiple Service Type=LoadBalancer
+implementations in a cluster, as different workloads may want to leverage different loadbalancer
+providers based on efficiency, availability, cost and other factors.
+
+For example, a cluster admin may want to use a public load balancer from a cloud provider
+for workloads that must be assigned a publically routable address, but they may want to
+enable a lower-cost solution for workloads that are only internally accessible.
```

Comment:
True, but it's still coupled to the cloud providers implementation.

Reply:
This is kinda of what I was referring to here https://github.com/kubernetes/enhancements/pull/1960#discussion_r480622402.

Since a bunch of providers already support their own "class", we probably don't want providers using the class annotation for varying classes still owned by the same cloud provider. The class should only apply if an entirely new implemenation/controller owns the resource.

Reply:
Also, 

> but they may want to enable a lower-cost solution for workloads that are only internally accessible.

This does not always mean using the "internal" version of the cloud provider's LB service, you may want to use something like MetalLB or your own solution.

Reply:
Example: AWS recently had to implement this for inside their in-tree cloud provider to allow an external controller to manage it as there was no generic way to tell the cloud provider not to act - https://github.com/kubernetes/kubernetes/pull/92839